data/routers/actionRouter.js

```diff
@@ -0,0 +1,96 @@
+const express = require('express');
+
+const router= express.Router();
+const actionDb = require('../helpers/actionModel');
+const projectDb = require('../helpers/projectModel');
+
+router.get('/', (req, res) => {
+  actionDb.get()
+    .then((actions) => {
+      res.status(200).json(actions)
+    })
+    .catch(error => {
+      res.status(500).json({message: "not found"})
+  
+    })
+})
+
+router.get('/:id', validatedActionId, (req, res) => {
+  res.status(200).json(req.action)
+});
+
+router.post('/', ValidatedAction, validateProjectId, (req, res) => {
+  actionDb.insert(req.body)
+    .then(newAction => {
+      res.status(201).json(newAction)
+    })
+    .catch(err => {
+      res.status(500).json({ message: "Error adding the project."})
+  }) 
+});
+
+router.put('/:id', ValidatedAction, validatedActionId, validateProjectId, (req, res) => {
+  const id = req.params.id
+  actionDb.update(id, req.body)
+    .then(updatedAction => {
+      res.status(200).json(updatedAction)
+    })
+    .catch(err => {
+      res.status(500).json({message: "Error updating the action."})
+    })
+})
+
+router.delete('/:id', validatedActionId, (req, res) => {
+  actionDb.remove(req.action.id)
+    .then(response => res.sendStatus(204))
+    .catch(err=> res.status(500).json({error: "The action with the id could not be retrieved."}))
+})
+
+function validatedActionId(req, res, next) {
+  const actionId = req.params.id;
+  if(!actionId) {
+    return res.status(404).json({message: `No action id`})
+  } 
+  actionDb.get(actionId)
+    .then(action => {
+       if (!action) {
+         return res.status(404).json({message: "No action id found."})
+       } else {
+         req.action = action;
+         next()
```

Comment:
Because we're querying the database, this 404 error message would be something closer to `"No action found with this id"`. We do have an id, but we just couldn't find an action that matches our id '' '

Reply:
You are right, I was in a hurry to complete the project. I will be more mindful.